### PR TITLE
Fine-grained: Fix AST merge issues

### PR DESCRIPTION
**Pull request by JukkaL**
This fixes all but one AST merge checker test failures. The remaining
one seems tricky to fix so I'll leave it for later.

I decided not to add tests for things that can be caught using the
merge checker. We'll have to run the merge checker periodically
since it's not enabled by default in the test suite.

### CODE REVIEW

File: mypy/types.py [hunk 1]
```diff
@@ -242,9 +242,10 @@ def serialize(self) -> JsonDict:
 class TypeList(Type):
     """Information about argument types and names [...].
 
-    This is only used for the arguments of a Callable type, i.e. for
+    This is used for the arguments of a Callable type, i.e. for
     [arg, ...] in Callable[[arg, ...], ret]. This is not a real type
-    but a syntactic AST construct.
+    but a syntactic AST construct. UnboundTypes can also have TypeList
+    types before they are processed into Callable type.s
```
Comment (msullivan) [inline mypy/types.py]: Typo

File: mypy/types.py [hunk 2]
```diff
@@ -1503,6 +1503,11 @@ def visit_type_type(self, t: TypeType) -> T:
     def visit_forwardref_type(self, t: ForwardRef) -> T:
         raise RuntimeError('Internal error: unresolved forward reference')
 
+    def visit_type_list(self, t: TypeList) -> T:
```
Comment (msullivan) [inline mypy/types.py]: Can other sorts of "synthetic" types show up in our traversal?

Would it be a better approach in the future to make the merger a synthetic type traverser?

Comment (JukkaL) [reply mypy/types.py]: Hmm looks like we could be hitting other such types as well. Maybe making it a "synthetic type visitor" is the right thing to do. I'll consider making a follow-up PR about this. 

On the other hand, we should perhaps only have one sort of type visitor. I don't really see the benefit of having two sorts of visitors.

**Comment by JukkaL** [comment]
The remaining failure is `testFineGrainedCallable`.

**Pull request merged**